nexus/reconfigurator/planning/src/blueprint_builder.rs

```diff
@@ -566,6 +721,88 @@ impl<'a> BlueprintBuilder<'a> {
+    ///
+    /// Returns the list of removed [`OmicronZoneConfig`] instances, or an
+    /// error if the sled ID wasn't found.
+    pub fn sled_expunge_all_zones(
+        &mut self,
+        sled_id: Uuid,
+    ) -> Result<Vec<OmicronZoneConfig>, Error> {
+        slog::debug!(self.log, "expunging all zones for sled"; "sled_id" => %sled_id);
+
+        self.sled_expunge_zones(sled_id, |_| true)
+    }
+
+    /// Expunges a set of zones corresponding to this sled ID.
+    ///
+    /// Returns the list of expunged [`OmicronZoneConfig`] instances, or an
+    /// error if the sled ID wasn't found.
+    ///
+    /// Return true in `expunge_fn` to remove a zone, or false to retain it.
+    pub fn sled_expunge_zones(
+        &mut self,
+        sled_id: Uuid,
+        mut expunge_fn: impl FnMut(&OmicronZoneConfig) -> bool,
+    ) -> Result<Vec<OmicronZoneConfig>, Error> {
+        // Check the sled id and return an appropriate error if it's invalid.
+        let _ = self.sled_resources(sled_id)?;
+
+        // Find all the zone IDs to remove in this sled.
+        let sled_zones = self.zones.current_sled_zones(sled_id);
+        let zone_ids_to_expunge = sled_zones
+            .filter(|z| expunge_fn(z))
+            .map(|z| z.id)
+            .collect::<BTreeSet<_>>();
+
+        slog::debug!(
+            self.log,
+            "expunging zones for sled";
+            "sled_id" => %sled_id,
+            "zone_ids_to_expunge" => ?zone_ids_to_expunge,
+        );
+
+        // Check that all the zone IDs exist. It is an internal invariant
+        // violation for a zone to be in the list of current sled zones, but
+        // not in zones_in_service.
+        if !zone_ids_to_expunge.is_subset(&self.zones_in_service) {
+            return Err(Error::Planner(anyhow!(
+                "zone IDs to remove are not a subset of zones in service: \
+                 zone_ids_to_remove: {:?}, zones_in_service: {:?}",
+                zone_ids_to_expunge,
+                self.zones_in_service,
+            )));
+        }
+
+        for z in &zone_ids_to_expunge {
+            self.zones_in_service.remove(z);
+        }
+
+        let sled_zones = self.zones.change_sled_zones(sled_id);
+        let zones = std::mem::replace(&mut sled_zones.zones, Vec::new());
+        let (expunge, keep) = zones
+            .into_iter()
+            .partition::<Vec<_>, _>(|z| zone_ids_to_expunge.contains(&z.id));
+        sled_zones.zones = keep;
+
+        // Also add the list of expunged Nexus zones to self.expunged_nexus_zones.
+        self.expunged_nexus_zones.extend(expunge.iter().filter_map(|z| {
+            (z.zone_type.tag() == ZoneTag::Nexus).then_some(z.id)
+        }));
+
+        Ok(expunge)
+    }
+
+    // ---
+    // Helper methods
+    // ---
+
+    // Precondition: we must have checked that it is valid for the zone to be
+    // added to the sled. (TODO: actually check this here, or use the type
+    // system to statically verify this.)
```

Comment:
Can you say more about this?  What does "valid" mean?  Does this panic or return an error or do something bad if this precondition is violated?